source/extensions/omni.isaac.lab/test/controllers/test_operational_space.py

```diff
@@ -0,0 +1,864 @@
+            else:
+                raise ValueError("Undefined target_type within _update_target().")
+
+        # update the target desired pose in world frame (for marker)
+        ee_target_pos_w, ee_target_quat_w = combine_frame_transforms(
+            root_pose_w[:, 0:3], root_pose_w[:, 3:7], ee_target_pose_b[:, 0:3], ee_target_pose_b[:, 3:7]
+        )
+        ee_target_pose_w = torch.cat([ee_target_pos_w, ee_target_quat_w], dim=-1)
+
+        next_goal_idx = (current_goal_idx + 1) % len(target_set)
+
+        return command, ee_target_pose_b, ee_target_pose_w, next_goal_idx
+
+    def _convert_to_task_frame(
+        self, osc: OperationalSpaceController, command: torch.tensor, ee_target_pose_b: torch.tensor
+    ):
+        command = command.clone()
+        task_frame_pose_b = None
+        if self.frame == "root":
+            # No need to transform anything if they are already in root frame
+            pass
+        elif self.frame == "task":
+            # Convert target commands from base to the task frame
+            command = command.clone()
+            task_frame_pose_b = ee_target_pose_b.clone()
+
+            cmd_idx = 0
+            for target_type in osc.cfg.target_types:
+                if target_type == "pose_abs":
+                    command[:, :3], command[:, 3:7] = subtract_frame_transforms(
+                        task_frame_pose_b[:, :3], task_frame_pose_b[:, 3:], command[:, :3], command[:, 3:7]
+                    )
+                    cmd_idx += 7
+                elif target_type == "pose_rel":
+                    # Compute rotation matrices
+                    R_task_b = matrix_from_quat(task_frame_pose_b[:, 3:])  # Task frame to base frame
+                    R_b_task = R_task_b.mT  # Base frame to task frame
+                    # Transform the delta position and orientation from base to task frame
+                    command[:, :3] = (R_b_task @ command[:, :3].unsqueeze(-1)).squeeze(-1)
+                    command[:, 3:7] = (R_b_task @ command[:, 3:7].unsqueeze(-1)).squeeze(-1)
+                    cmd_idx += 6
+                elif target_type == "wrench_abs":
+                    # These are already defined in target frame for ee_goal_wrench_set_tilted_task (since it is
+                    # easier), so not transforming
+                    cmd_idx += 6
+                else:
+                    raise ValueError("Undefined target_type within _convert_to_task_frame().")
+        else:
+            # Raise error for invalid frame
+            raise ValueError("Invalid frame selection for target setting inside the test_operational_space.")
+
+        return command, task_frame_pose_b
+
+    def _check_convergence(
+        self,
+        osc: OperationalSpaceController,
+        ee_pose_b: torch.tensor,
+        ee_target_pose_b: torch.tensor,
+        ee_force_b: torch.tensor,
+        ee_target_b: torch.tensor,
+    ):
```

Comment:
please add docstrings for this function

Reply:
"""Check the convergence to the target.

        Args:
            osc (OperationalSpaceController): The operational space controller.
            ee_pose_b (torch.tensor): The end-effector pose in the body frame.
            ee_target_pose_b (torch.tensor): The end-effector target pose in the body frame.
            ee_force_b (torch.tensor): The end-effector force in the body frame.
            ee_target_b (torch.tensor): The end-effector target in the body frame.

        Raises:
            AssertionError: If the convergence is not achieved.
            ValueError: If the target type is undefined.
        """